### PR TITLE
[06] Chunked cross-entropy for large vocab

### PR DESCRIPTION
**Pull request by xjdr-noumena**
## Summary

Chunked linear cross-entropy to avoid materializing full logits tensor `[N, vocab]`. For 200k+ vocab, this saves ~26GB+ in bf16.

- Token chunking (2048) + vocab blocking (4096) 
- 2-pass streaming softmax: logsumexp → probs → grads
- No `[N, V]` allocation ever created

## Changes

- `nmoe/loss.py` (new) - chunked_cross_entropy + memcheck CLI
- `nmoe/train.py` - use return_hidden path
- `nmoe/model.py` - add return_hidden=True support
- `tests/test_loss_chunked.py` - parity tests

## PR Attestation

```yaml
nmoe_pr:
  issue: 06
  baseline_sha: 0fee6e5
  risk: warm

gates_passed:
  tier_a:
    - cpu:syntax
  tier_b: []

perf:
  gate: n/a
  notes: Memory reduction feature, not throughput
```

## Test plan

- [ ] Tier A: syntax check
- [ ] Tier B: hot_path_min (moonlet_1x10, moonlight_8x20)
- [ ] Memory check: `python -m nmoe.loss` shows savings vs dense

### CODE REVIEW

**Comment by xjdr-noumena** [comment]
## Blocker: 13x performance regression

Profiled on debug pod with moonlet shapes (N=32768, D=2048, V=201088, bf16):

| Path | Forward | Backward | Total |
|------|---------|----------|-------|
| Chunked (chunk_vocab=4096) | 119ms | 1297ms | **1417ms** |
| Dense F.cross_entropy | 33ms | 74ms | **107ms** |

**Root causes (from nsys profiler):**
1. Backward GEMMs running as fp32 SGEMM (~79% of CUDA time)
2. `torch.nonzero` in inner loop triggers cudaStreamSync (~879ms CPU stall)

**Gap needs to be addressed prior to merge.** Options:
- CUDA kernel for fused chunked CE (like flash-attention approach)
- Triton kernel with proper bf16 tensor core usage
- liger-kernel or similar existing implementation

**Comment by xjdr-noumena** [comment]
Parked: moved to quack implementation in master
